### PR TITLE
fix(interpreter): deepen the String-exotic index predicate into string_exotic_index

### PR DESCRIPTION
**Pull request by pmatos**
## What this is

An **architecture audit** (`improve-codebase-architecture` skill) surfaced a scattered spec predicate; this PR concentrates it into one deep module, driven **test-first** (`tdd` skill). It is a *deepening* refactor in the house style of #371 (`deepen @@toStringTag`) and 96ac1b0 (`deepen StringToNumber; concentrate the WhiteSpace predicate`) — and, like the latter, the concentration exposes and fixes real ECMAScript divergences.

## Reviewing the audit results — the candidates

The scan turned up three deepening opportunities. Ranked on the task's criteria (impact on maintainability + real technical debt, safety, testability):

| Candidate | Shape | Real bug? | Sites | Verdict |
|---|---|---|---|---|
| **String-exotic index predicate** | §10.4.3 own-index test reimplemented via `str::parse::<usize>()` | **Yes** — 5 confirmed divergences | **11** + a missing Reflect check | ✅ **chosen** |
| ToLength (§7.1.20) | `min(2^53-1)` open-coded, no helper | No | ~11 | clean concentration, no bug |
| BigInt↔Number ordering | mixed-comparison block copied 5× in `eval.rs` | No | 5 | pure, testable, no bug |

The string-index candidate wins: strongest **locality** payoff (one predicate, eleven callers that had already drifted), a real bug surface, and it fits an established repo pattern — the canonical helper `canonical_numeric_index_string` already existed and was simply not used on these paths.

## The friction

Eleven String-exotic MOP paths each re-derived *"is P an own character index?"* with `str::parse::<usize>()` + a code-unit length compare. Rust's integer parser accepts spellings that are **not** CanonicalNumericIndexStrings (a leading `+`, leading zeros), so a look-alike key was treated as a real indexed property. The interface was as complex as the implementation, copied eleven times — a shallow, drift-prone seam.

## The deepening

One predicate, `types::string_exotic_index(key, len) -> Option<usize>` (CanonicalNumericIndexString + IsIntegralNumber + reject `-0` + range), placed next to `canonical_numeric_index_string`. Every String-exotic operation — `[[Get]]`, `[[GetOwnProperty]]`, `[[HasProperty]]`, `[[DefineOwnProperty]]`, `[[Set]]`, `[[Delete]]` — now crosses that single seam.

## Divergences fixed (each cross-checked with node)

- `"abc"["01"]` / `"abc"["+1"]` returned `"b"` → now `undefined`.
- `"01" in Object("abc")`, `getOwnPropertyDescriptor(Object("abc"),"01")`, and strict `delete Object("abc")["01"]` all treated `"01"` as an own index → now ordinary (absent / deletable).
- `Object.defineProperty(Object("abc"), "01", {...})` threw a `TypeError` from the string-index redefinition guard → now definable.
- `proxy_delete_property` measured length with `to_string().chars().count()` (code **points**), so `Reflect.deleteProperty(Object("💩"),"1")` returned `true` → now `false` (UTF-16 code **units**).
- `Reflect.deleteProperty` had **no** String-exotic `[[Delete]]` check at all → added via the new helper.
- Bonus: the read fast paths now delegate out-of-range keys to the prototype walk, fixing `"abc"[5]` missing `String.prototype[5]`.

**Scope:** concentrates the String-exotic index predicate only. Array / arguments / typed-array index paths keep their own canonical rules and are untouched.

## Tests validating the refactor (test-first, red → green)

- **Unit** — `string_exotic_index_tests` (types.rs): the ECMA-262 truth table — canonical/non-canonical spellings, `-0`, non-integral, `Infinity`/`NaN`, range. (Red before the helper existed.)
- **JS seam** — `string_exotic_index_seam_tests` (tests.rs): Get / `in` / GetOwnProperty / defineProperty / strict+sloppy delete / Reflect / non-BMP. (Red on the old predicate.)
- **test262-extra** — `String-exotic-canonical-index.js`: negative edge cases test262 doesn't cover, using the standard `assert`/`verifyProperty` harness.

## Verification (run to completion, this branch)

- `cargo test --release`: **352 passed, 0 failed** + smoke oracle ok.
- test262, **0 regressions**: `built-ins/String` 2443/2443, `built-ins/Object` 6802/6802, `built-ins/Reflect` + `language/expressions/delete` + `language/expressions/property-accessors` 451/451.
- test262-extra new file 2/2; custom tests 10/10; `for-in` / spread / `Object.assign` enumeration byte-identical to node.
- `./scripts/lint.sh` clean (rustfmt + clippy).

🤖 Generated with Claude Code

### CODE REVIEW

File: src/interpreter/eval.rs [hunk 1]
```diff
@@ -1428,14 +1432,10 @@ impl Interpreter {
...
-                        ])))
-                    } else {
-                        Completion::Normal(JsValue::Undefined)
-                    }
+                } else if let Some(idx) =
+                    crate::interpreter::types::string_exotic_index(&name, s.code_units.len())
+                {
+                    Completion::Normal(JsValue::String(JsString::from_vec(vec![s.code_units[idx]])))
                 } else if let Some(sp_id) = self.realm().string_prototype {
                     Completion::Normal(self.get_property_on_id(sp_id, &name))
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/interpreter/eval.rs:1440]: **Out of scope for this PR** — Confirmed real and reproduced exactly as described (`"abc"["01"]` → `42`, `"abc"?.["01"]` → `undefined` on jsse vs. `42`/`42` on Node, with `String.prototype["01"]` defined as an accessor). Traced to `access_property_on_value` in `src/interpreter/eval.rs` falling back to `get_property_on_id` (which never invokes getters) for the optional-chain path, while regular member access correctly uses `get_object_property` (which does).

However, the defective line is pre-existing and unchanged by this PR (`git show 8e79153` shows only the predicate check above it changed), and the same bug generalizes to Number/Boolean/Symbol/BigInt optional-chain access — it's a general "optional chaining on primitives skips prototype accessor getters" defect, orthogonal to string-exotic-index concentration, which is this PR's stated sole scope.

Tracked as #384 (https://github.com/pmatos/jsse/issues/384).

_This assessment was made by two independent AI reviewers (Claude Opus 4.6 and Codex GPT-5.6). If you disagree, please reply and we'll re-evaluate._
